license_manager/proto/license_manager.proto

```diff
@@ -0,0 +1,171 @@
+syntax = "proto3";
+
+package license_manager.proto;
+
+import "google/protobuf/timestamp.proto";
+
+// Describes a vendor-agnostic frontend for allocating licenses to hardware tool
+// command invocations.
+// Terminology:
+//
+//   License: A tuple of (vendor, license_type) strings that identifies a
+//     resource that can only be used by one process at a time. These resources
+//     are ultimately fetched from and returned to a vendor-specific license
+//     server.
+//
+//     Licenses are "allocated" when the LicenseManager frontend associates it
+//     with an invocation.
+//
+//     Licenses are "in use" when the invocation actually uses a license from
+//     the underlying license server. Licenses should not be "in use" unless
+//     they are also "allocated", but "allocated" licenses are not always "in
+//     use".
+//
+//   Invocation: A single execution of a particular hardware toolchain binary.
+//     Invocations may use a license, and to do so they must be allocated one
+//     before the underlying hardware toolchain process is allowed to execute.
+
+service LicenseManager {
+  // Allocate attempts to allocate a specific license type for this invocation.
+  // If successful, invocations will keep the allocation alive by calling
+  // Refresh(). If the invocation is queued for the license instead, invocations
+  // will continue polling with Allocate() until allocation is successful.
+  rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+
+  // Refresh refreshes a license allocation while an invocation is still
+  // potentially using the license. If the invocation fails to refresh the
+  // allocation, the underlying license may be allocated to another invocation.
+  rpc Refresh(RefreshRequest) returns (RefreshResponse) {}
+
+  // LicensesStatus returns the status of all license types, as reported by both
+  // the LicenseManager and the underlying license servers.
+  rpc LicensesStatus(LicensesStatusRequest) returns (LicensesStatusResponse) {}
```

Comment:
this is good for debugging and CLI, although as discussed the primary monitoring will be via prometheus.

Reply:
My thoughts exactly.